File: cli/internal/scheme/scheme.go
// Package scheme defines the abstraction each payload version of the txtshr
// specification (SPEC.md §4) implements. Concrete versions (v0 raw, v1
// PBKDF2+AES-GCM, …) each satisfy Scheme and are registered in Registry,
// keyed by their version identifier ("0", "1", …).
//
// This layer is deliberately independent of the wire format: Params carries
// raw bytes under the key names each version's section of §4 defines, and
// translating them to and from the URL-fragment encoding (SPEC.md §3 —
// key=value syntax, base64url) is the fragment layer's concern.
package scheme

// Params holds a payload's version-specific parameters: raw bytes keyed by
// the names the version's section of SPEC.md §4 defines (e.g. v1 uses "s",
// "n", "c"). Each version owns its key vocabulary; the keys "v" and "r" are
// reserved by SPEC.md §3 and must not be used. The version identifier is not
// part of Params — it lives alongside the Scheme's registration.
type Params map[string][]byte

// Scheme is one version of the txtshr payload scheme: the pair of Encode and
// Decode operations from SPEC.md §2, specialized by a §4 version section.
type Scheme interface {
	// Encrypts reports whether this version encrypts the plaintext.
	// For versions that return false, the passphrase arguments below are
	// ignored, and the caller is responsible for obtaining the user's
	// explicit per-invocation acknowledgement before emitting a payload
	// (SPEC.md §4.0). For versions that return true, callers must supply a
	// non-empty passphrase.
	Encrypts() bool

	// Encode produces the version-specific payload parameters for the given
	// plaintext, per the version's section of SPEC.md §4.
	Encode(plaintext []byte, passphrase string) (Params, error)

	// Decode recovers the plaintext from payload parameters, mirroring
	// Encode. Per SPEC.md §2.2 it must fail — not return garbage — when
	// parameters the version requires are missing or malformed and, for
	// encrypting versions, when the passphrase is wrong or the payload has
	// been tampered with. Unknown keys are ignored (SPEC.md §3).
	Decode(p Params, passphrase string) ([]byte, error)
}

File: viewer/src/utils/scheme.ts
// The viewer-side seam for txtshr payload versions, mirroring the CLI's
// internal/scheme package: each version of SPEC.md §4 implements Scheme
// (v0.ts, v1.ts, …) and is registered in registry.ts, keyed by its version
// identifier ("0", "1", …). This layer deals in raw bytes; translating the
// URL fragment's key=value + base64url wire format (SPEC.md §3) into Params
// is the caller's concern.

/**
 * A payload's version-specific parameters: raw bytes keyed by the names the
 * version's section of SPEC.md §4 defines (e.g. v1 uses "s", "n", "c"). The
 * keys "v" and "r" are reserved by SPEC.md §3 and never appear here.
 */
export type Params = Record<string, Uint8Array>;

/**
 * The decryption half of one version of the txtshr payload scheme — the
 * viewer only ever consumes payloads, so Encrypt (SPEC.md §2.1) has no
 * counterpart here.
 */
export interface Scheme {
  /**
   * Whether this version encrypts the plaintext. Drives the passphrase
   * prompt and, when false, the unencrypted-content indicator (SPEC.md §4.0).
   */
  readonly encrypts: boolean;
  /** Params keys this version's section of SPEC.md §4 requires. */
  readonly requiredParams: readonly string[];
  /**
   * Recovers the plaintext from payload parameters (SPEC.md §2.2). Throws —
   * never returns garbage — when required params are malformed and, for
   * encrypting versions, when the passphrase is wrong or the payload has
   * been tampered with. Unknown keys are ignored (SPEC.md §3).
   */
  decode(params: Params, passphrase: string): Promise<string>;
}
